platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,482 @@
+                .max(Instant::compareTo)
+                .orElseGet(Instant::now);
+        snapshotCompactionLock.acquire();
+        try {
+            currentCompactionStartTime.set(startTime);
+            newCompactedFiles.clear();
+            startNewCompactionFile(targetCompactionLevel);
+        } finally {
+            snapshotCompactionLock.release();
+        }
+
+        // We need a map to find readers by file index below. It doesn't have to be synchronized
+        // as it will be accessed in this thread only, so it can be a simple HashMap or alike.
+        // However, standard Java maps can only work with Integer, not int (yet), so auto-boxing
+        // will put significant load on GC. Let's do something different
+        int minFileIndex = Integer.MAX_VALUE;
+        int maxFileIndex = 0;
+        for (final DataFileReader<?> r : filesToCompact) {
+            minFileIndex = Math.min(minFileIndex, r.getIndex());
+            maxFileIndex = Math.max(maxFileIndex, r.getIndex());
+        }
+        final int firstIndexInc = minFileIndex;
+        final int lastIndexExc = maxFileIndex + 1;
+        final DataFileReader<?>[] readers = new DataFileReader[lastIndexExc - firstIndexInc];
+        for (DataFileReader<?> r : filesToCompact) {
+            readers[r.getIndex() - firstIndexInc] = r;
+        }
+
+        boolean allDataItemsProcessed = false;
+        try {
+            final KeyRange keyRange = dataFileCollection.getValidKeyRange();
+            index.forEach((path, dataLocation) -> {
+                if (!keyRange.withinRange(path)) {
+                    return;
+                }
+                final int fileIndex = DataFileCommon.fileIndexFromDataLocation(dataLocation);
+                if ((fileIndex < firstIndexInc) || (fileIndex >= lastIndexExc)) {
+                    return;
+                }
+                final DataFileReader<?> reader = readers[fileIndex - firstIndexInc];
+                if (reader == null) {
+                    return;
+                }
+                final long fileOffset = DataFileCommon.byteOffsetFromDataLocation(dataLocation);
+                // Take the lock. If a snapshot is started in a different thread, this call
+                // will block until the snapshot is done. The current file will be flushed,
+                // and current data file writer and reader will point to a new file
+                snapshotCompactionLock.acquire();
+                try {
+                    final DataFileWriter<?> newFileWriter = currentWriter.get();
+                    long serializationVersion = reader.getMetadata().getSerializationVersion();
+                    final long newLocation = newFileWriter.writeCopiedDataItem(
+                            serializationVersion, reader.readDataItemBytes(fileOffset));
+                    // update the index
+                    index.putIfEqual(path, dataLocation, newLocation);
+
```

Comment:
As far as I understand, the main goal was to achieve simplicity and predictability here. Since it's processing data locations one by one, it can quickly interrupt the process if it comes to that. Also, using atomic updates it's not interfering with the transaction processing. 

Yes, I see your point - if it processed these data items in chunks, we could reduce some overhead. For example, we wouldn't have to acquire a lock for every single entry (and there may be billions there). However, it necessary to be extremely careful here to make sure that these optimizations don't affect the primary use of MerkleDB. 

I'd say, this work is beyond the scope of particular PR.

Reply:
Make sense.  I wasn't making a point, I was just verifying that it is *intended* to serialize the process.
It's easy to end up unintentionally serializing things when using a Semaphore.